nomad/leader.go

```diff
@@ -2766,18 +2767,32 @@ func (s *Server) initializeKeyring(stopCh <-chan struct{}) {
 		return
 	}
 
-	err = s.encrypter.AddKey(rootKey)
+	isClusterUpgraded := ServersMeetMinimumVersion(
+		s.serf.Members(), s.Region(), minVersionKeyringInRaft, true)
+
+	wrappedKeys, err := s.encrypter.AddUnwrappedKey(rootKey, isClusterUpgraded)
 	if err != nil {
 		logger.Error("could not add initial key to keyring", "error", err)
 		return
 	}
-
-	if _, _, err = s.raftApply(structs.RootKeyMetaUpsertRequestType,
-		structs.KeyringUpdateRootKeyMetaRequest{
-			RootKeyMeta: rootKey.Meta,
-		}); err != nil {
-		logger.Error("could not initialize keyring", "error", err)
-		return
+	if isClusterUpgraded {
+		logger.Warn("cluster is upgraded to 1.9.0: initializing keyring")
+		if _, _, err = s.raftApply(structs.WrappedRootKeysUpsertRequestType,
+			structs.KeyringUpsertWrappedRootKeyRequest{
+				WrappedRootKeys: wrappedKeys,
+			}); err != nil {
+			logger.Error("could not initialize keyring", "error", err)
+			return
+		}
+	} else {
+		logger.Warn("cluster is not upgraded to 1.9.0: initializing legacy keyring")
```

Comment:
> I don't even know how people would hit it?!

It's possible to bootstrap a new cluster with mixed versions, if you have poor deployment practices or otherwise like to cause yourself pain. Seems incredibly unlikely, but I wanted to make sure we were covering our bases.